journal.py

```diff
@@ -91,43 +93,49 @@ def __repr__(self):
 # all the views
 
 
-@view_config(route_name='add', request_method='POST')
-def add_entry(request):
-    title = request.params.get('title')
-    text = request.params.get('text')
-    Entry.write(title=title, text=text)
-    return HTTPFound(request.route_url('home'))
-
-
-@view_config(route_name='add', renderer="templates/add.jinja2")
+@view_config(route_name='add', xhr=True, renderer='json')
+@view_config(route_name='add', xhr=False, renderer="templates/add.jinja2")
 def add_entry_view(request):
-    username = request.params.get('username', '')
-    return {'username': username}
+    if not request.authenticated_userid:
+        return HTTPFound(request.route_url('login'))
 
+    if request.method == 'POST':
+        title = request.params.get('title')
+        text = request.params.get('text')
+        Entry.write(title=title, text=text)
+        return HTTPFound(request.route_url('home'))
```

Comment:
Again, if the request was made via ajax, this is not going to do what you really want it to do.  You should be returning the new post's data as a JSON blob, rather than issuing a redirect.